index.src.html

```diff
@@ -1841,73 +1849,269 @@
 </dl>
 
 
-Raw Media Interfaces (Frames){#raw-media-interfaces}
+Raw Media Interfaces {#raw-media-interfaces}
 ====================================================
 These interfaces represent unencoded (raw) media.
 
+Memory Model {#raw-media-memory-model}
+--------------------------------------
+
+### Background ### {#raw-media-memory-model-background}
 
-AudioFrame Interface {#audioframe-interface}
+This section is non-normative.
+
+Decoded media data may occupy a large amount of system memory. To minimize the
+need for expensive copies, this specification defines a scheme for reference
+counting (`clone()` and `close()`).
+
+### Reference Counting ### {#raw-media-memory-model-reference-counting}
+
+A <dfn>media resource</dfn> is storage for the actual pixel data or the audio
+sample data described by a {{VideoFrame}} or {{AudioData}}.
+
+The {{AudioData}} {{AudioData/[[resource reference]]}} and {{VideoFrame}}
+{{VideoFrame/[[resource reference]]}} internal slots hold a reference to a
+[=media resource=].
+
+{{VideoFrame}}.{{VideoFrame/clone()}} and
+{{AudioData}}.{{AudioData/clone()}} return new frame objects who's
+`[[resource reference]]` points to the same [=media resource=] as the original
+frame.
+
+{{VideoFrame}}.{{VideoFrame/close()}} and {{AudioData}}.{{AudioData/close()}}
+will clear their [[resource reference]] slot, releasing the reference their
+[=media resource=]
+
+A [=media resource=] must remain alive at least as long as it continues to be
+referenced by a `[[resource reference]]`.
+
+NOTE: When a [=media resource=] is no longer referenced by a
+    `[[resource reference]]`, the resource may be destroyed. User agents are
+    encouraged to destroy such resources quickly to reduce memory pressure and
+    facilitate resouce reuse.
+
+AudioData Interface {#audiodata-interface}
 ---------------------------------------------
 
 <xmp class='idl'>
 [Exposed=(Window,DedicatedWorker)]
-interface AudioFrame {
-  constructor(AudioFrameInit init);
+interface AudioData {
+  constructor(AudioDataInit init);
+
+  readonly attribute AudioSampleFormat sampleFormat;
+  readonly attribute float sampleRate;
+  readonly attribute unsigned long numberOfFrames;
+  readonly attribute unsigned long numberOfChannels;
+  readonly attribute unsigned long allocationSize;
+  readonly attribute unsigned long long duration;
   readonly attribute unsigned long long timestamp;
-  readonly attribute AudioBuffer? buffer;
+
+  undefined copyTo([AllowShared] BufferSource destination, unsigned long planeNumber);
```

Comment:
I'm thinking maybe an offset and a frame count would make it a lot easier to send data to the WASM heap, and would help to address [this feedback](https://github.com/w3c/webcodecs/issues/176) from game developers (while not exactly that if I remember our conversation, it's really close).

After demuxing properly (with all the edge cases, easier said than done!), you have quite often a decoder delay value in frames, and a padding value in frames (is WAV the _only_ codec that doesn't have this? Certainly among the codecs available on the Web). Then imagine you're doing all your work in a custom audio engine written in C++, so the immediate thing you're doing is to copy the audio frames out to the WASM heap.

Say you're doing AAC, commonly it's 2112 frames of padding (not aligned to the common number of 1024 frames per packets), so you want to drop the two first packet entirely, and then copy to your heap with an offset of 64 frames starting on the third packet. Similarly, for the last packet, you only need to copy part of the packet, that's a function of a duration you find in the mp4 container usually (not always).

Here with the current API, you need to copy out to some intermediary buffer, and then copy from an offset to end final location, which is wasteful.

I believe adding two parameters

```
 undefined copyTo([AllowShared] BufferSource destination, unsigned long planeNumber, offset = 0, frameCount = 0);
```

with following meaning:
- `offset` is an offset in frames in the plane - throw `RangeError` if bigger than this `AudioData`'s `numberOfFrames`
- `frameCount` is the number of frames to copy - throw `RangeError` if bigger than `numberOfFrames - offset` or `destination.length`. The Web Audio API clamps in a similar situation, but I'd rather be explicit in this situation. The special value of 0 can be decided to mean `min(numberOfFrames - offset, destination.length)` if we want, that's ergonomic.

Most of the time you're not using those parameters though, hence the meaning of the defautl values.

Reply:
I'm not a fan of APIs which take an index like `planeNumber'. It sounds instead like we should have a planes array on the AudioData object, so folks can say data.planes[i].copyTo(dest, offset, frameCount). This would provide some symmetry with the VideoFrame API as well. This also makes it clear that you're not simply indexing into a channel array.

Reply:
I'm not a fan of APIs which take an index like `planeNumber'. It sounds instead like we should have a planes array on the AudioData object, so folks can say data.planes[i].copyTo(dest, offset, frameCount). This would provide some symmetry with the VideoFrame API as well. This also makes it clear that you're not simply indexing into a channel array.

Reply:
@padenot , re offset + framecount + RangeErrors: all SGTM. Out of time tonight, but I'll send a new commit ASAP. I vote we create a `AudioDataReadToOptions` dict that lists these members alongside a required member for `planeNumber`. This also should help w/ @dalecurtis's concern that the simple argument looks like a channel index. 

Re: `planes` symmetry... not sure on that. We may yet delete `planes` from VideoFrame ([still up in the air](https://github.com/w3c/webcodecs/issues/157#issuecomment-832214060), and adding a new Plane interface with a single copyTo() method feels like a lot of ceremony IMO.

Reply:
WFM, thanks!

Reply:
https://github.com/w3c/webcodecs/issues/223 opened for followup, linking to the valuable informations in this thread.